### PR TITLE
Support expansion of nested macros

### PR DESCRIPTION
**Pull request by ahoppen**
The basic idea is that a `sourcekit-lsp://swift-macro-expansion` URL should have sufficient information to reconstruct the contents of that macro buffer without relying on any state in SourceKit-LSP. The benefit of not having any cross-request state in SourceKit-LSP is that an editor might can send the `workspace/getReferenceDocument` request at any time and it will succeed independent of the previous requests. Furthermore, we can always get the contents of the macro expansion to form a `DocumentSnapshot`, which can be used to provide semantic functionality inside macro expansion buffers.

To do that, the `sourcekit-lsp:` URL scheme was changed to have a parent instead of a `primary`, which is the URI of the document that the buffer was expanded from. For nested macro expansions, this will be a `sourcekit-lsp://swift-macro-expansion` URL itself.

With that parent, we can reconstruct the macro expansion chain all the way from the primary source file. To avoid sending the same expand macro request to sourcekitd all the time, we introduce `MacroExpansionManager`, which caches the last 10 macro expansions.

`SwiftLanguageService` now has a `latestSnapshot` method that returns the contents of the reference document when asked for a reference document URL and only consults the document manager for other URIs. To support semantic functionality in macro expansion buffers, we need to call that `latestSnapshot` method so we have a document snapshot of the macro expansion buffer for position conversions and pass the following to the sourcekitd requests.
```
keys.sourceFile: snapshot.uri.sourcekitdSourceFile,
keys.primaryFile: snapshot.uri.primaryFile?.pseudoPath,
```

We should consider if there’s a way to make the `latestSnapshot` method on `documentManager` less accessible so that the method which also returns snapshots for reference documents is the one being used by default.

CC @lokesh-tr

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

**Review by lokesh-tr — APPROVED** [review]
I love how you implemented it and am glad that my initial idea of having nested URLs is good and that the macro expansions storage idea took its shape into a Cache system which is pretty neat.

Thank you so much for lending me an helping hand. ❤️ 

I will look into building semantic functionality on top of this and see if there exists any issues in integrating this with VS Code / Extension.

**Comment by ahoppen** [comment]
@hamishknight @bnbarham If you have review comments, I will address them in a follow-up PR so that @lokesh-tr can continue building semantic functionality in macro expansion buffers on top of this.

File: Sources/SourceKitLSP/Swift/MacroExpansionReferenceDocumentURLData.swift [hunk 1]
```diff
@@ -15,41 +15,71 @@ import LanguageServerProtocol
...
+  /// URL: `file:///path/to/swift_file.swift`
+  /// ```swift
+  /// let a = 10
+  /// let b = 5
+  /// print(#stringify(a + b))
+  /// ```
+  ///
+  /// Generated content of reference document url:
+  /// URL:
+  /// `sourcekit-lsp://swift-macro-expansion/L3C7-L3C23.swift?primaryFilePath=/path/to/swift_file.swift&fromLine=3&fromColumn=8&toLine=3&toColumn=8&bufferName=@__swift_macro_..._Stringify_.swift`
```
Comment (lokesh-tr) [inline Sources/SourceKitLSP/Swift/MacroExpansionReferenceDocumentURLData.swift:50]: I do notice `primaryFilePath` should be `parent` in doc comment similarly in a few other places. I shall update this in my PR.

File: Sources/SourceKitLSP/Swift/MacroExpansion.swift
Comment (lokesh-tr) [inline Sources/SourceKitLSP/Swift/MacroExpansion.swift:1]:
```
let req = PeekDocumentsRequest(
          uri: expandMacroCommand.textDocument.uri,
          position: expandMacroCommand.positionRange.lowerBound,
          locations: expansionURIs
        )
```

I guess the uri and position are pointing to the reference document here. It should actually point to the primaryFile.

I shall update this in the new PR.

**Review by lokesh-tr — COMMENTED** [review]
Just noticed a few things while working on the new PR, I will fix them.

**Comment by lokesh-tr** [comment]
Here's the new one: https://github.com/swiftlang/sourcekit-lsp/pull/1634

File: Sources/SourceKitLSP/Swift/MacroExpansion.swift
Comment (ahoppen) [reply Sources/SourceKitLSP/Swift/MacroExpansion.swift:1]: Thank you for addressing these in your PR @lokesh-tr! Appreciate it.

File: Sources/SourceKitLSP/Swift/ReferenceDocumentURL.swift [hunk 2]
```diff
@@ -95,6 +102,32 @@ package enum ReferenceDocumentURL {
...
+      self.pseudoPath
+    }
+  }
+
+  /// If this is a URI to a reference document, the URI of the source file from which this reference document was
+  /// derived.
+  ///
+  /// The primary file is used to determine the workspace and language service that is used to generate the reference
+  /// document as well as getting the reference document's build settings.
+  var primaryFile: DocumentURI? {
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SourceKitLSP/Swift/ReferenceDocumentURL.swift:123]: The usages that will be introduced by https://github.com/swiftlang/sourcekit-lsp/pull/1634 will expect `primaryFile` to be `nil` if the URI is an actual on-disk file, so I think keeping it optional makes sense.

File: Sources/SourceKitLSP/Swift/ExpandMacroCommand.swift [hunk 3]
```diff
@@ -13,9 +13,7 @@
 import LanguageServerProtocol
 import SourceKitD
 
-package struct ExpandMacroCommand: RefactorCommand {
-  typealias Response = MacroExpansion
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SourceKitLSP/Swift/ExpandMacroCommand.swift:17]: Yes, thanks for catching.

File: Sources/SourceKitLSP/Swift/MacroExpansion.swift [hunk 4]
```diff
@@ -62,23 +199,30 @@ extension SwiftLanguageService {
       throw ResponseError.unknown("Connection to the editor closed")
     }
 
-    guard let primaryFileURL = expandMacroCommand.textDocument.uri.fileURL else {
-      throw ResponseError.unknown("Given URI is not a file URL")
-    }
+    let primaryFileDisplayName =
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SourceKitLSP/Swift/MacroExpansion.swift:202]: Oh yes, good suggestion.

File: Sources/SourceKitLSP/Swift/SwiftLanguageService.swift [hunk 5]
```diff
@@ -373,32 +391,37 @@ extension SwiftLanguageService {
...
-      try documentManager.latestSnapshot(notification.textDocument.uri)
-    }
-    guard let snapshot else {
-      return
-    }
-    cancelInFlightPublishDiagnosticsTask(for: snapshot.uri)
-    await diagnosticReportManager.removeItemsFromCache(with: snapshot.uri)
+    switch try? ReferenceDocumentURL(from: notification.textDocument.uri) {
+    case .macroExpansion:
+      break
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SourceKitLSP/Swift/SwiftLanguageService.swift:396]: I want to migrate generated interfaces to reference documents next, which will allow us to provide semantic functionality in them. They will require logic for open/close/reopen, so I don’t think that the early return scales. That’s why I would just keep it as a switch like it is today.

**Pull request merged**